### PR TITLE
Matchmaking/Battle contract integration

### PR DESCRIPTION
**Pull request by benzo-diazepine**
Summary of changes:
- Implemented matchmaking
  - Enter with deck
  - Exit
  - Start battle once another player enters and has a compatible ranking
  - Show battle scene when the battle event is detected
- Adapted game logic to match the PepemonBattle logic
  - Battles are now consistent between the blockchain and the game client
- Updated some definitions to match the contract's properties (Cards properties more specifically)

Autogenerated files:
Assets/Scripts/Contracts/PepemonMatchmaker/abi/ContractDefinition/PepemonMatchmakerDefinition.cs

Future improvements:
Check if player's deck is in the wait list when accessing the Matchmaker screen, in cases where the game crashed and the player had to restart it, this would give the player the chance to remove the deck from the wait list.

### CODE REVIEW

File: Assets/Scripts/Controllers/BattlePrepController.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    {
+        Debug.Log("Received battle event. BattleId: " + battleEventData.BattleId);
+        Debug.Log("Loading battle data..");
+
+        var reqBattleRngSeed = PepemonBattle.GetBattleRNGSeed(battleEventData.BattleId);
+        var reqPlayer1BattleCard = PepemonCardDeck.GetBattleCard(battleEventData.Player1Deck);
+        var reqPlayer2BattleCard = PepemonCardDeck.GetBattleCard(battleEventData.Player2Deck);
+        var reqPlayer1SupportCards = PepemonCardDeck.GetAllSupportCards(battleEventData.Player1Deck);
+        var reqPlayer2SupportCards = PepemonCardDeck.GetAllSupportCards(battleEventData.Player2Deck);
+        await Task.WhenAll(reqBattleRngSeed, reqPlayer1BattleCard, reqPlayer2BattleCard, reqPlayer1SupportCards, reqPlayer2SupportCards);
```
Comment (0xdenkle) [inline Assets/Scripts/Controllers/BattlePrepController.cs:120]: There's a window here where the battle has been created but the exit button can still be pressed.
I'd recommend disabling input on the exit button after the BattleCreated event is picked up

File: Assets/Scripts/Controllers/GameController.cs [hunk 2]
```diff
@@ -91,11 +129,34 @@ IEnumerator StartRound()
...
+            //Need to refresh decks
+
+            // Shuffle player1 support cards
+            _player1.GetAndShuffelDeck(69, _roundNumber, battleSeed);
+
+            //Reset played card count
+            _player1.PlayedCardCount = 0;
+
+            // Shuffle player2 support cards
+            _player2.GetAndShuffelDeck(420, _roundNumber, battleSeed);
```
Comment (0xdenkle) [inline Assets/Scripts/Controllers/GameController.cs:148]: Would be good to see constants for these magic numbers (69, 420, 69420)

**Review by 0xdenkle — CHANGES_REQUESTED** [review]
Nice, pretty much just those two small problems above. Code is really clean.

File: Assets/Scripts/Controllers/GameController.cs (hunk 2 above)
Comment (benzo-diazepine) [reply Assets/Scripts/Controllers/GameController.cs:148]: Nice. Okay

File: Assets/Scripts/Controllers/BattlePrepController.cs (hunk 1 above)
Comment (benzo-diazepine) [reply Assets/Scripts/Controllers/BattlePrepController.cs:124]: Updated in [this ](https://github.com/pepemon-dao/pepemon-battle-degen/pull/48/commits/60b768f7be47f29f9a22f5590f431c19447d0f3a) commit

File: Assets/Scripts/Controllers/GameController.cs (hunk 2 above)
Comment (benzo-diazepine) [reply Assets/Scripts/Controllers/GameController.cs]: Updated in [this](https://github.com/pepemon-dao/pepemon-battle-degen/pull/48/commits/60b768f7be47f29f9a22f5590f431c19447d0f3a) commit

**Review by 0xdenkle — APPROVED** [review]

**Review by 0xdenkle — APPROVED** [review]

**Pull request merged**